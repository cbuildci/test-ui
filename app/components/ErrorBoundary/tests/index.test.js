import React from 'react';
// import { shallow } from 'enzyme';
import renderer from 'react-test-renderer';

import ErrorBoundaryWithConsumer, { ErrorContextProvider, ErrorBoundary } from '../index';

describe('ErrorBoundary (inner)', () => {
    it('should render children if no error', () => {
        const renderedComponent = renderer
            .create((
                <ErrorBoundary>
                    <h1>Test</h1>
                </ErrorBoundary>
            ))
            .toJSON();

        expect(renderedComponent).toMatchSnapshot();
    });

    it('should catch errors and show nothing by default', async () => {
        const mockCallback = jest.fn();
        const err = new Error('foobar');

        function FailingComponent() {
            throw err;
        }

        await new Promise((resolve) => {
            function onError(...args) {
                mockCallback(...args);
                resolve();
            }

            const renderedComponent = renderer
                .create((
                    <ErrorBoundary onError={onError}>
                        <FailingComponent/>
                    </ErrorBoundary>
                ))
                .toJSON();

            expect(renderedComponent).toMatchSnapshot();
        });

        expect(mockCallback.mock.calls.length).toBe(1);
        expect(mockCallback.mock.calls[0].length).toBe(2);
        expect(mockCallback.mock.calls[0][0]).toBe(err);
        expect(mockCallback.mock.calls[0][1]).toEqual({
            componentStack: '\n    in FailingComponent\n    in ErrorBoundary',
        });
    });

    it('should catch errors and show the error message (as component)', () => {
        function FailingComponent() {
            throw new Error('foobar');
        }

        const renderedComponent = renderer
            .create((
                <ErrorBoundary errorMessage={<div>Error</div>}>
                    <FailingComponent/>
                </ErrorBoundary>
            ))
            .toJSON();

        expect(renderedComponent).toMatchSnapshot();
    });

    it('should catch errors and show the error message (as function)', () => {
        const err = new Error('foobar');

        function FailingComponent() {
            throw err;
        }

        const mockErrorMessage = jest.fn(() => <div>Error</div>);

        const renderedComponent = renderer
            .create((
                <ErrorBoundary errorMessage={mockErrorMessage}>
                    <FailingComponent/>
                </ErrorBoundary>
            ))
            .toJSON();

        expect(renderedComponent).toMatchSnapshot();
        expect(mockErrorMessage.mock.calls.length).toBe(1);
        expect(mockErrorMessage.mock.calls[0].length).toBe(2);
        expect(mockErrorMessage.mock.calls[0][1]).toEqual(
            expect.objectContaining({
                componentStack: expect.any(String),
            }),
        );
    });
});

describe('ErrorContextProvider (default export) with ErrorContextProvider', () => {
    it('should call error callback provided to context provider', async () => {
        const mockCallback = jest.fn();
        const err = new Error('foobar');

        function FailingComponent() {
            throw err;
        }

        await new Promise((resolve) => {
            function onError(...args) {
                mockCallback(...args);
                resolve();
            }

            const renderedComponent = renderer
                .create((
                    <ErrorContextProvider value={onError}>
                        <ErrorBoundaryWithConsumer>
                            <FailingComponent/>
                        </ErrorBoundaryWithConsumer>
                    </ErrorContextProvider>
                ))
                .toJSON();

            expect(renderedComponent).toMatchSnapshot();
        });

        expect(mockCallback.mock.calls.length).toBe(1);
        expect(mockCallback.mock.calls[0].length).toBe(2);
        expect(mockCallback.mock.calls[0][0]).toBe(err);
        expect(mockCallback.mock.calls[0][1]).toEqual({
            componentStack: '\n    in FailingComponent\n    in ErrorBoundary\n    in ErrorBoundaryWithContext',
        });
    });
});
