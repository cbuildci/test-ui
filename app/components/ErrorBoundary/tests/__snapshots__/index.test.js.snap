// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ErrorBoundary (inner) should catch errors and show nothing by default 1`] = `null`;

exports[`ErrorBoundary (inner) should catch errors and show the error message (as component) 1`] = `
<div>
  Error
</div>
`;

exports[`ErrorBoundary (inner) should catch errors and show the error message (as function) 1`] = `
<div>
  Error
</div>
`;

exports[`ErrorBoundary (inner) should render children if no error 1`] = `
<h1>
  Test
</h1>
`;

exports[`ErrorContextProvider (default export) with ErrorContextProvider should call error callback provided to context provider 1`] = `null`;
